src/herder/HerderPersistenceImpl.cpp

```diff
@@ -93,6 +93,13 @@ HerderPersistenceImpl::saveSCPHistory(uint32_t seq,
         }
     }
 
+    if (!qmapOptional)
```

Comment:
this is wrong: not only you're skipping saving the qmap (what you were trying to do here - but that we should not do), but you're also not saving qset information from the messages that are included in the SCP dump

Reply:
Discussed offline -- right now persistence logic assumes that we always save the latest state, which might not be the case anymore. The solution is to make changes to sql queries to make sure we're not overriding newer info.